mlir/lib/Mitigation/Transforms/MitigationMethods/Zne.cpp

```diff
@@ -34,33 +35,142 @@
+    return fnFoldedOp;
+}
+
 LogicalResult ZneLowering::match(mitigation::ZneOp op) const { return success(); }
 
+// TODO: Optimize the traversal of call graphs (currently used twice)
+// Also all functions exploree in the call graph get their ZNE version.
 void ZneLowering::rewrite(mitigation::ZneOp op, PatternRewriter &rewriter) const
 {
     Location loc = op.getLoc();
-
+    auto moduleOp = op->getParentOfType<ModuleOp>();
     // Number of folds
     auto numFolds = op.getNumFolds();
     RankedTensorType numFoldType = cast<RankedTensorType>(numFolds.getType());
     const auto sizeInt = numFoldType.getDimSize(0);
 
     // Folding type
     auto foldingAlgorithm = op.getFolding();
-
-    // Create the folded circuit function
-    FlatSymbolRefAttr foldedCircuitRefAttr =
-        getOrInsertFoldedCircuit(loc, rewriter, op, foldingAlgorithm);
-    func::FuncOp foldedCircuit =
-        SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, foldedCircuitRefAttr);
-
+    auto calleeOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(op, op.getCalleeAttr());
+
+    func::FuncOp fnFoldedOp;
+
+    if (!calleeOp->hasAttr("qnode")) {
+        // Traverse the callgraph, copy all the function to a `.zne` version and fold qnodes
+        traverseCallGraph(calleeOp, /*symbolTable=*/nullptr, [&](func::FuncOp funcOp) {
+            if (!funcOp->hasAttr("qnode")) {
+                PatternRewriter::InsertionGuard insertGuard(rewriter);
+                rewriter.setInsertionPointToStart(moduleOp.getBody());
+
+                // Copy the function and create a .zne counter part and add the scale factor as last
+                // argument
+                auto currentFnFoldedOp = createZneFunc(funcOp, rewriter);
+                // Folding of the qnodes and replace their calls with the folded version
+                if (containsQnodes(currentFnFoldedOp)) {
+                    currentFnFoldedOp.walk([&](func::CallOp callOp) {
+                        PatternRewriter::InsertionGuard insertGuard(rewriter);
+                        func::FuncOp funcOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                            op, callOp.getCalleeAttr());
+                        std::string foldedName = callOp.getCalleeAttrName().str() + ".folded";
+                        func::FuncOp foldedOp = SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                            moduleOp, rewriter.getStringAttr(foldedName));
+                        if (funcOp->hasAttr("qnode") and !foldedOp) {
+                            // Create the folded circuit function
+                            auto foldedCircuitAttr =
+                                getOrInsertFoldedCircuit(loc, rewriter, funcOp, foldingAlgorithm);
+                            func::FuncOp foldedOp =
+                                SymbolTable::lookupNearestSymbolFrom<func::FuncOp>(
+                                    moduleOp, foldedCircuitAttr);
```

Comment:
Would it be cleaner for `getOrInsertFoldedCircuit` to return the FuncOp directly instead of the symbol? It looks like this function + symbol lookup is used in a few places.

Reply:
I prefer not to change it as I need to change all the folding functions for that